SI-1786 incorporate defined bounds in inference

paulp opened the pull request:
Also fixes SI-5459.

Look, you don't have to redeclare the bounds, isn't it exciting? For instance,
there are eight places in JavaMirrors with this: `jTypeVariable[_ <: GenericDeclaration]` After this code is in starr, those can look like this: `jTypeVariable[_]`. Since TypeVariable's definition looks like this, `interface TypeVariable<D extends GenericDeclaration>`, we already know that!

paulp committed:
SI-1786 incorporate defined bounds in inference

Also fixes SI-5459.

Look, you don't have to redeclare the bounds, isn't it exciting?
For instance, there are eight places in JavaMirrors with this:

  jTypeVariable[_ <: GenericDeclaration]

After this code is in starr, those can look like this:

  jTypeVariable[_]

Since TypeVariable's definition looks like this:

  interface TypeVariable<D extends GenericDeclaration>

We already know that!

paulp commented:
Review by @retronym or @adriaanm

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -4867,19 +4867,23 @@ trait Typers extends Adaptations with Tags {
+        val tpt        = tree.tpt
+        val args       = tree.args
+        val tpt1       = typed1(tpt, mode | FUNmode | TAPPmode, WildcardType)
+        def isPoly     = tpt1.tpe.isInstanceOf[PolyType]
```
We could probably have this as a method in `Types.scala`?

paulp replied:
Indeed, we may as well collect the whole set.

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -4888,19 +4892,39 @@ trait Typers extends Adaptations with Tags {
-              case _ =>
-            })
+            foreach2(args, tparams) { (arg, tparam) =>
+              // note: can't use args1 in selector, because Binds got replaced
```
Is the omission of the apostrophe intentional?

paulp replied:
Yes, because I read it as "the Bind nodes were replaced", but now I guess it means "the Bind's symbol was replaced", yes?

xeno-by replied:
Looks like that yeah, though I'm not an expert here.

On Sun, May 12, 2013 at 11:07 PM, Paul Phillips notifications@github.com
wrote:

> > @@ -4888,19 +4892,39 @@ trait Typers extends Adaptations with Tags {
> >                }
> >              val argtypes = args1 map (_.tpe)
> > -            foreach2(args, tparams)((arg, tparam) => arg match {
> > -              // note: can't use args1 in selector, because Bind's got replaced
> > -              case Bind(_, _) =>
> > -                if (arg.symbol.isAbstractType)
> > -                  arg.symbol setInfo // XXX, feedback. don't trackSymInfo here!
> > -                    TypeBounds(
> > -                      lub(List(arg.symbol.info.bounds.lo, tparam.info.bounds.lo.subst(tparams, argtypes))),
> > -                      glb(List(arg.symbol.info.bounds.hi, tparam.info.bounds.hi.subst(tparams, argtypes))))
> > -              case _ =>
> > -            })
> > -            foreach2(args, tparams) { (arg, tparam) =>
> > -              // note: can't use args1 in selector, because Binds got replaced
> >   Yes, because I read it as "the Bind nodes were replaced", but now I guess it means "the Bind's symbol was replaced", yes?
> >   ---
> >   Reply to this email directly or view it on GitHub:
> >   https://github.com/scala/scala/pull/2518/files#r4186481

xeno-by commented on test/files/pos/t1786.scala:
```diff
@@ -0,0 +1,19 @@
+class A {
+  def f1(i:MyClass[_])                       = i.myValue.intValue
+  def f2(i:MyClass[_ <: SomeClass])          = i.myValue.intValue
+  def f3[T](i: MyClass[T])                   = i.myValue.intValue
```
Even though `f1` feels intuitive, this one is odd. Personally I would expect `T <: SomeClass` to be required to compile.

paulp replied:
Well, think of it this way.

```
abstract class A {
  type Foo <: String
  val s: Foo
}
class B {
  // Should I be able to write this?
  def f1(x: A) = x.s.length
  // Or should I have to write this?
  def f2(x: A { type Foo <: String }) = x.s.length
}
```

xeno-by replied:
But `f3` is different in a sense, isn't it? There we have type information propagate from a method's parameter to a method's type parameter, which is something new to Scala, right? Note that I'm not against the idea, I'm just trying to understand it better!

Oh, also. Which of those are compilation errors?

```
class SomeClass
class MyClass[T <: SomeClass] 
class C[T] {
  def x: MyClass[T] = ???
}
new C[Int]
new C[SomeClass]
```

paulp replied:
Just as one would hope:
`./a.scala:9: error: type arguments [Int] do not conform to class C's type parameter bounds [T <: SomeClass]
  new C[Int]
      ^
one error found`

The restrictions are identical; the difference is in whether said restrictions are to be ignored at the point when they are most needed.

I can see that it would be nice if the error message said something more like

`...do not conform to class C's (inferred from def x: MyClass[T]) type parameter bounds [T <: SomeClass]`

xeno-by replied:
I see. That's consistent.

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -4888,19 +4892,39 @@ trait Typers extends Adaptations with Tags {
+                  && tparam.hasCompleteInfo /* SI-2940 */
+                  && !tparam.isFBounded     /* SI-2251 */
+                  && !tparam.isHigherOrderTypeParameter
+                  && !(abounds.hi <:< tbounds.hi)
```
Since the check includes calls to `hasCompleteInfo`, I have a question. How do we know that type applications won't change behavior if we shuffle declaration order or file compilation order?

paulp replied:
I can't prove it, but by lots of observation, it comes through here twice, and it's only the first pass where it doesn't have complete info. I will gladly receive insight on this one.

paulp committed:
No bounds-driven inference for the named.

This is a variation on the previous commit which excludes
named type parameters, so this works

  def f(x: Class[_]) = x.foo.bar

But this does not:

  def f[T](x: Class[T]) = x.foo.bar

This seems undesirable to me, but I offer it in case it makes
the patch more attractive to others.

paulp commented:
I pushed a variation which only works on underscores and not on named tparams.

adriaanm commented:
It is much more attractive to me to restrict the bound inference to synthetic symbols. It would be weird to suddenly bestow bounds on type parameters that weren't declared.

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -4888,19 +4892,40 @@ trait Typers extends Adaptations with Tags {
+                if (!(lo =:= lo0 && hi =:= hi0))
+                  asym setInfo logResult(s"Updating bounds of ${asym.fullLocationString} in $tree from '$abounds' to")(TypeBounds(lo, hi))
+              }
+              if (asym != null && asym.isAbstractType) {
```
why not `&& canEnhanceIdent`, or else include it in the guard for `case Bind` ?

paulp replied:
Because those conditions didn't exist on Bind before and I was afraid of unintended consequences.

adriaanm commented:
LGTM since it's only acting on _
